### PR TITLE
Allow transferring datetime2/datetimeoffset to/from SQL Server 2012

### PR DESCRIPTION
**Pull request by ztzg**
This is quite a hack, but enables round-tripping full-precision timestamps to SQL Server despite the fact that Mono's current TDS implementation is limited to protocol version 7.1.

Please cf. second commit message for details.

### CODE REVIEW

**Comment by tritao** [comment]
Can you provide a test case?

File: mcs/class/Mono.Data.Tds/Mono.Data.Tds.Protocol/Tds70.cs [hunk 1]
```diff
@@ -586,6 +586,13 @@ private void WriteParameterInfo (TdsMetaParameter param)
 			} else if (colType == TdsColumnType.BigVarBinary) {
 				if (size > 8000)
 					colType = TdsColumnType.Image;
+			} else if (colType == TdsColumnType.DateTime2 ||
+				   colType == TdsColumnType.DateTimeOffset) {
+				// HACK: Wire-level DateTime{2,Offset}
+				// require TDS 7.3, which this driver
+				// does not implement correctly--so we
+				// serialize to ASCII instead.
```
Comment (atsushieno) [inline mcs/class/Mono.Data.Tds/Mono.Data.Tds.Protocol/Tds70.cs]: out of curiousity, do you know what's not implemented correctly?

**Comment by atsushieno** [comment]
Merged all those SqlClient PRs. Thanks!

File: mcs/class/Mono.Data.Tds/Mono.Data.Tds.Protocol/Tds70.cs (hunk 1 above)
Comment (ztzg) [inline mcs/class/Mono.Data.Tds/Mono.Data.Tds.Protocol/Tds70.cs]: Hi Atsushi,

I haven't investigated fully, but have written my conclusions in the
commit message: https://github.com/ztzg/mono/commit/f055c46a

**Comment by buraktamturk** [comment]
I have a problem with mapping DateTime: http://stackoverflow.com/questions/31499849/mono-invalidoperationexception-on-datetime-column-sql-server

I wonder whatever this commit has anything to do with it. (e.g. a step closer for fixing it or worse?) 

I really would love to contribute on Mono project. I just need a root cause of the problem.

**Comment by ztzg** [comment]
Hi Burak,

Without looking at the details, I would say a step closer, as Mono's TDS driver did not know about `DateTime2` at all before my changes.  You may want to try and see if your code happens to work (possibly with a loss of precision) on a version of Mono which does not include my patches.

As I have commented here:

  https://github.com/mono/mono/pull/1771#issuecomment-114206833

the `DateTime2` features (and the TDS driver, in general) are known not to be complete nor flawless.

An ideal solution would be to integrate MS' "referencesource" implementation with Mono, but that is easier said than done as it depends on low-level Windows APIs.

The missing functionality can of course also be implemented in the current driver, but I would expect a fair amount of work to be necessary for complex `DT2`-using .NET applications to work out of the box on Mono.

HTH, -D

**Comment by buraktamturk** [comment]
Wow, just now, I notice @EvK is also have this problem like me in #1771 

However I was having mapping issues too, but after this commit(s) there is a cast issue. So this is, for sure, we are very close, thanks @ztzg 

The root cause could be: EF *might* think that TDS has returned string instead of DateTime.

**Comment by EvK** [comment]
@buraktamturk for now I use the following trick (if you are using Entity Framework): open .edmx as xml file, change ProviderManifestToken attribute in Schema element from 2008(2012) to 2005. That way EF will think you are working with SQL Server 2005 and will not use DateTime2 everywhere. Obviously that will work if you don't use modern column types like "geomerty" etc. However, for me so far it helps.

**Comment by buraktamturk** [comment]
@EvK Much thanks for the response, I read it earlier but I had to temporarily switch Windows for development. With the new version of Mono (4.2), I just wanted to back on my os x.

Actually, I do not have the edmx file, I wrote models and dbcontext myself. So I researched code configuration instead of xml configuration. I still get same exception however I think the following code works replacing the "ProviderManifestToken" as it says "hah1" as intended. I found it in internet.

I think it does not force Mono to use SQL 2005's protocol. Is there any way force Mono to use SQL 2005's protocol instead of using newer one?

```
internal sealed class EntityFrameworkDbConfiguration : DbConfiguration
  {
      /// <summary>
      /// The provider manifest token to use for SQL Server.
      /// </summary>
      private const string SqlServerManifestToken = @"2005";

      /// <summary>
      /// Initializes a new instance of the <see cref="EntityFrameworkDbConfiguration"/> class.
      /// </summary>
      public EntityFrameworkDbConfiguration()
      {

          this.AddDependencyResolver(new SingletonDependencyResolver<IManifestTokenResolver>(new ManifestTokenService()));
            Console.WriteLine("asd");
        }

      /// <inheritdoc />
      private sealed class ManifestTokenService : IManifestTokenResolver
      {
          /// <summary>
          /// The default token resolver.
          /// </summary>
          private static readonly IManifestTokenResolver DefaultManifestTokenResolver = new DefaultManifestTokenResolver();

          /// <inheritdoc />
          public string ResolveManifestToken(DbConnection connection)
          {
              if (connection is SqlConnection)
              {
                    Console.WriteLine("hah 1");
                    return SqlServerManifestToken;
              }

Console.WriteLine("hah 2");
              return DefaultManifestTokenResolver.ResolveManifestToken(connection);
          }
      }
  }

[DbConfigurationType(typeof(EntityFrameworkDbConfiguration))]
public class CourseContext : DbContext
{
    // ...
}
```

**Comment by buraktamturk** [comment]
I place Console.WriteLine to ProcessColumnInfo function in TDS80 class as ```Console.WriteLine(col.ColumnName + " " + col.ColumnType);```

It shows DateTime2 columns as NVarChar. Should I look the culprit there? Or the shortly, how does Mono's implementation gives the row to the application? I could try string to datetime convertion before giving the data (object) to the EF (if the col name contains "date"), to see whatever it works or not, I just need to know which file I should look.

**Comment by buraktamturk** [comment]
Furthermore, I found which file to look (Tds.cs), TdsColumnType is sent 231 for the DateTime2 values as well as NVarChar. How I can know this column is a DateTime?

**Comment by buraktamturk** [comment]
I think TDS version 0x71 advertise DateTime columns as NVarChar, setting it 0x73 on client side causes connection to drop as this driver does not implements 0x73 as mentioned in this PR. 

However, since I only develop on Mono and not targeting Mono on release. I found a temporary solution that works for me. I hope it will help someone here.

Since SQL server tell the column as NVarChar (code 231) in TDS 0x71, I couldn't find a more reliable solution to check whatever it is DateTime or not, I had to check whatever the column (string in here) contains date by trying converting it into DateTime. I also had to ignore the exception given here because the column might not be a date but a 27 character length string.

Here is the modified case in GetColumnValue function in Tds class in Tds.cs file: 

**this is just the demonstration of the problem, should not get merged**

```csharp
case TdsColumnType.NVarChar :
enc = GetEncodingFromColumnCollation (lcid, sortId);				
if (outParam) 
	comm.Skip (1);
element = GetStringValue (colType, true, outParam, enc);

try {
	if(element != null && ((string)element).Length == 27 ) {
			element = DateTime.ParseExact((string) element, "yyyy-MM-dd HH:mm:ss.fffffff", System.Globalization.CultureInfo.InvariantCulture);
	}
} catch(Exception e) {

}

break;
```

**Pull request merged**